### ISSUE TITLE
EmbeddingBag is too big causing error "assertion failed: byte_count_ < total_size_"

### ISSUE THREAD

**Issue by shz0116** [pytorch/xla#2371]
I am trying to initialize an embedding table with size EmbeddingBag(30000000, 128). And it caused the above assertion error. This is  related with the 2GB Protocol Buffer limit. (also partially discussed in closed issue [#1633](https://github.com/pytorch/xla/issues/1633) before). I was asked to open a new issue here so that we can directly trace the progress.

**Comment by taylanbil** [comment]
@davidel , would it be possible to send the large tensor in smaller chunks and then stitch on the device?

**Comment by davidel** [comment]
Everything is possible, but that is a pretty complex change, thrown away in a few months (with 1VM).

**Comment by taylanbil** [comment]
Here is a workaround:

```python
import torch
import torch_xla.core.xla_model as xm
import torch.nn as nn

x = nn.EmbeddingBag(30000000, 128)
a = torch.split(x.weight, x.weight.size(0)//10, dim=0)                                                                            a = list(a)
d = xm.xla_device()

print('sending chunks to device', flush=True)
for i, chunk in enumerate(a):
    print('\tchunk', i)
    a[i] = chunk.to(d)

t = nn.Parameter(torch.ones(10,10))
x.weight = t
x = x.to(d)
print('sent x to device', flush=True)
a = torch.cat(a)
x.weight = nn.Parameter(a)
print(x.weight.shape)
```

output:

```
sending chunks to device
        chunk 0
        chunk 1
        chunk 2
        chunk 3
        chunk 4
        chunk 5
        chunk 6
        chunk 7
        chunk 8
        chunk 9
sent x to device
torch.Size([30000000, 128])
```

**Comment by davidel** [comment]
Oh, yeah. I thought you meant it do it within the core C++ layers (which required both client and server side changes).

**Comment by taylanbil** [comment]
Yeah actually that was what I asked (c++ level change) :) Since that is not happening per your comment, I put together the python side workaround..

@shz0116 , closing this issue, please re-open if this doesn't work.

**Comment by shz0116** [comment]
Tried the code. It runs ok. However, I got OOM problem when we actually run the lookups. It 
Attach the following code:

batch_size = 4
nnz = 10
indices  = torch.randint(0, num_features, (batch_size, nnz)).to(d)
results = x(indices)

0 derived errors ignored.
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch_xla/__init__.py", line 57, in _prepare_to_exit
    _XLAC._prepare_to_exit()
RuntimeError: Resource exhausted: From /job:tpu_worker/replica:0/task:0:
2 root error(s) found.
  (0) Resource exhausted: Failed to allocate request for 14.30GiB (15360000000B) on device ordinal 0
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

	 [[XRTExecute_G12]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

  (1) Resource exhausted: Failed to allocate request for 14.30GiB (15360000000B) on device ordinal 0
	 [[{{node XRTExecute}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

===========================================
If I shrink the size to 6000000, I got a different error:
sending chunks to device
	chunk 0
	chunk 1
	chunk 2
	chunk 3
	chunk 4
	chunk 5
	chunk 6
	chunk 7
	chunk 8
	chunk 9
sent x to device
torch.Size([6000000, 128])
Finished generating indices
2020-07-25 00:37:23.812175: W   20851 tensorflow/core/distributed_runtime/rpc/grpc_remote_master.cc:160] RPC failed with status = "Unavailable: Socket closed" and grpc_error_string = "{"created":"@1595637443.812045947","description":"Error received from peer ipv4:10.136.249.194:8470","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC
Traceback (most recent call last):
  File "splitemb.py", line 33, in <module>
    results = x(indices)
  File "/anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch/nn/modules/module.py", line 726, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch/nn/modules/sparse.py", line 307, in forward
    per_sample_weights, self.include_last_offset)
  File "/anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch/nn/functional.py", line 2007, in embedding_bag
    include_last_offset)
RuntimeError: tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:439 : Check failed: session->session()->Run( session_work->feed_inputs, session_work->outputs_handles, &outputs) == ::tensorflow::Status::OK() (Aborted: Session bbf408e8ad292df7 is not found. vs. OK)
*** Begin stack trace ***
	tensorflow::CurrentStackTrace[abi:cxx11]()

	clone
*** End stack trace ***

terminate called after throwing an instance of 'std::runtime_error'
  what():  tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1104 : Check failed: session->session()->Run( feed_inputs, {}, {cached_node.operations[0]}, &outputs) == ::tensorflow::Status::OK() (Aborted: Session be9eeb1be4e80290 is not found. vs. OK)
*** Begin stack trace ***
	tensorflow::CurrentStackTrace[abi:cxx11]()
	xla::XrtComputationClient::ReleaseHandles(std::vector<xla::XrtComputationClient::DeviceHandle, std::allocator<xla::XrtComputationClient::DeviceHandle> >*, std::function<xla::XrtSession::CachedNode const& (xla::XrtSession*, tensorflow::Scope const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> const&, xla::metrics::Metric*, xla::metrics::Counter*)
	xla::XrtComputationClient::HandleReleaser()
	xla::util::TriggeredTask::Runner()
	
	clone
*** End stack trace ***

Aborted

**Comment by davidel** [comment]
`30000000 * 128 * 4 = 15.3GB`

Considering the reserved size, the TPU binary, etc.. that won't fit.
Or even if it fit in the trivial example, you won't be able to run much on the device.
